jetty-core/jetty-websocket/jetty-websocket-core-common/src/main/java/org/eclipse/jetty/websocket/core/internal/FrameFlusher.java

```diff
@@ -511,27 +518,62 @@ public String toString()
         return String.format("%s[queueSize=%d,aggregate=%s]",
             super.toString(),
             getQueueSize(),
-            batchBuffer);
+            _batchBuffer);
     }
 
-    private static class Entry extends FrameEntry
+    private class Entry extends FrameEntry implements CyclicTimeouts.Expirable
     {
-        private final long timeOfCreation = System.currentTimeMillis();
+        private long _expiry = Long.MAX_VALUE;
 
         private Entry(Frame frame, Callback callback, boolean batch)
         {
             super(frame, callback, batch);
+
+            long currentTime = NanoTime.now();
+            if (_frameTimeout > 0)
+            {
+                long frameExpiry = Math.addExact(currentTime, Duration.ofMillis(_frameTimeout).toNanos());
+                _expiry = nanoTimeMin(_expiry, frameExpiry);
+            }
+            if (_messageTimeout > 0)
+            {
+                if (frame.isDataFrame())
+                {
+                    // If this is the first frame of the message remember the message timeout.
+                    if (frame.getOpCode() != OpCode.CONTINUATION)
+                        _currentMessageTimeout = Math.addExact(currentTime, Duration.ofMillis(_messageTimeout).toNanos());
+
+                    _expiry = nanoTimeMin(_expiry, _currentMessageTimeout);
```

Comment:
Do we need to do this:
```suggestion
                    _expiry = nanoTimeMin(_expiry, _currentMessageTimeout);
                    
                    if (frame.isFin())
                        _currentMessageTimeout = 0;
```

Reply:
We don't need to. 

Every time a message starts it will be a frame with a TEXT or BINARY opcode and we set the `_currentMessageTimeout`, and the only place we use it is right here.